### PR TITLE
Show extended error information for ICPR

### PR DESCRIPTION
**Pull request by zeroSteiner**
Requires rapid7/windows_error#5

This adds extended error information to the ICPR library code which is used by the icpr_cert module and the new certifried exploit. When the certificate fails to issue for some reason, it looks up the error result as an [HRESULT](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-erref/0642cb2f-2075-4469-918c-4441e69c548a) and prints the details for the user to assist in troubleshooting.

Before these changes, the extended error information was simply missing.

## Verification

- [x] Follow the setup steps from #16939, skipping the "Configure a certificate template to be vulnerable via ESC1" section
- [x] Run the module as a real user but try to issue a certificate as a machine (which you are not)
    - [x] Set SMBUser / SMBPass to the credentials of the real user
    - [x] Set CERT_TEMPLATE to `Machine` to trigger the failure and see the error details
- [x] See detailed error messages saying that the use can not enroll in that type of certificate

## Example

```
msf6 auxiliary(admin/dcerpc/icpr_cert) > run
[*] Running module against 192.168.159.10

[*] 192.168.159.10:445 - Connecting to ICertPassage (ICPR) Remote Protocol
[*] 192.168.159.10:445 - Binding to \cert...
[+] 192.168.159.10:445 - Bound to \cert
[*] 192.168.159.10:445 - Requesting a certificate for user aliddle - digest algorithm: SHA256 - template: Machine
[-] 192.168.159.10:445 - There was an error while requesting the certificate.
[-] 192.168.159.10:445 - Denied by Policy Module
[-] 192.168.159.10:445 - Error details:
[-] 192.168.159.10:445 -   Source:  (0x0009) FACILITY_SECURITY: The source of the error code is the Security API layer.
[-] 192.168.159.10:445 -   HRESULT: (0x80094012) CERTSRV_E_TEMPLATE_DENIED: The permissions on the certificate template do not allow the current user to enroll for this type of certificate.
[*] Auxiliary module execution completed
msf6 auxiliary(admin/dcerpc/icpr_cert) > 
```

### CODE REVIEW

File: lib/msf/core/exploit/remote/ms_samr.rb [hunk 1]
```diff
@@ -108,9 +108,11 @@ def add_computer(opts = {})
   rescue RubySMB::Dcerpc::Error::SamrError => e
     raise MsSamrUnknownError, "A DCERPC SAMR error occured: #{e.message}"
   ensure
-    samr_con.samr.close_handle(user_handle) if user_handle
-    samr_con.samr.close_handle(samr_con.domain_handle) if samr_con.domain_handle
-    samr_con.samr.close_handle(samr_con.server_handle) if samr_con.server_handle
+    if samr_con
```
Comment (zeroSteiner) [inline lib/msf/core/exploit/remote/ms_samr.rb:111]: This bonus fix addresses a stack trace in the samr_computer module that would occur if the `SMBDomain` argument was completely invalid.

Before:

```
msf6 auxiliary(admin/dcerpc/samr_computer) > show options 

Module options (auxiliary/admin/dcerpc/samr_computer):

   Name               Current Setting  Required  Description
   ----               ---------------  --------  -----------
   COMPUTER_NAME                       no        The computer name
   COMPUTER_PASSWORD                   no        The password for the new computer
   RHOSTS             192.168.159.10   yes       The target host(s), see https://github.com/rapid7/metasploit-framework/wiki/Using-Metasploit
   RPORT              445              yes       The target port (TCP)
   SMBDomain          foo.foo          no        The Windows domain to use for authentication
   SMBPass            Password1!       no        The password for the specified username
   SMBUser            smcintyre        no        The username to authenticate as


Auxiliary action:

   Name          Description
   ----          -----------
   ADD_COMPUTER  Add a computer account



View the full module info with the info, or info -d command.

msf6 auxiliary(admin/dcerpc/samr_computer) > run
[*] Running module against 192.168.159.10

[-] 192.168.159.10:445 - Auxiliary failed: NoMethodError undefined method `domain_handle' for nil:NilClass
[-] 192.168.159.10:445 - Call stack:
[-] 192.168.159.10:445 -   /home/smcintyre/Repositories/metasploit-framework/lib/msf/core/exploit/remote/ms_samr.rb:112:in `ensure in add_computer'
[-] 192.168.159.10:445 -   /home/smcintyre/Repositories/metasploit-framework/lib/msf/core/exploit/remote/ms_samr.rb:113:in `add_computer'
[-] 192.168.159.10:445 -   /home/smcintyre/Repositories/metasploit-framework/modules/auxiliary/admin/dcerpc/samr_computer.rb:70:in `action_add_computer'
[-] 192.168.159.10:445 -   /home/smcintyre/Repositories/metasploit-framework/modules/auxiliary/admin/dcerpc/samr_computer.rb:54:in `run'
[*] Auxiliary module execution completed
msf6 auxiliary(admin/dcerpc/samr_computer) > run SMBDomain=msflab
[*] Running module against 192.168.159.10

[+] 192.168.159.10:445 - Successfully created msflab\DESKTOP-KATWN3WT$
[+] 192.168.159.10:445 -   Password: wXHEsDw2ohkFnhudb5RJFVJ6wYfMZMbX
[+] 192.168.159.10:445 -   SID:      S-1-5-21-3402587289-1488798532-3618296993-2121
[*] Auxiliary module execution completed
msf6 auxiliary(admin/dcerpc/samr_computer) > 
```

After:
```
msf6 auxiliary(admin/dcerpc/samr_computer) > run
[*] Running module against 192.168.159.10

[-] 192.168.159.10:445 - Auxiliary aborted due to failure: unexpected-reply: Error returned during domain lookup in SAM server: (0xc00000df) STATUS_NO_SUCH_DOMAIN: The specified domain did not exist.
[*] Auxiliary module execution completed
msf6 auxiliary(admin/dcerpc/samr_computer) >
```

**Comment by bwatters-r7** [comment]
I took myself off these today in case someone wants to grab them in a different time zone.  If they're still here tomorrow, I'll test/land.

**Review by cdelafuente-r7 — APPROVED** [review]

**Comment by bwatters-r7** [comment]
Hrm.....
I landed the windows_error stuff, built the new gem (0.1.5), verified the proper version, and I'm getting a NameError:
```
[*] 10.5.134.159:445 - Connecting to ICertPassage (ICPR) Remote Protocol
[*] 10.5.134.159:445 - Binding to \cert...
[+] 10.5.134.159:445 - Bound to \cert
[*] 10.5.134.159:445 - Requesting a certificate for user Administrator - digest algorithm: SHA256 - template: Machine
[-] 10.5.134.159:445 - There was an error while requesting the certificate.
[-] 10.5.134.159:445 - Denied by Policy Module
[-] 10.5.134.159:445 - Auxiliary failed: NameError uninitialized constant WindowsError::HResult
[-] 10.5.134.159:445 - Call stack:
[-] 10.5.134.159:445 -   /home/tmoose/rapid7/metasploit-framework/lib/msf/core/exploit/remote/ms_icpr.rb:201:in `do_request_cert'
[-] 10.5.134.159:445 -   /home/tmoose/rapid7/metasploit-framework/lib/msf/core/exploit/remote/ms_icpr.rb:89:in `request_certificate'
[-] 10.5.134.159:445 -   /home/tmoose/rapid7/metasploit-framework/modules/auxiliary/admin/dcerpc/icpr_cert.rb:64:in `action_request_cert'
[-] 10.5.134.159:445 -   /home/tmoose/rapid7/metasploit-framework/modules/auxiliary/admin/dcerpc/icpr_cert.rb:50:in `run'
[*] Auxiliary module execution completed

```
I'm using the new gem:
```
bundle info windows_error
Your /home/tmoose/.bundle/config config includes `BUNDLE_LOCAL__REX-POWERSHELL`, which contains the dash character (`-`).
This is deprecated, because configuration through `ENV` should be possible, but `ENV` keys cannot include dashes.
Please edit /home/tmoose/.bundle/config and replace any dashes in configuration keys with a triple underscore (`___`).
  * windows_error (0.1.5)
	Summary: Provides a way to look up Windows NTSTATUS and Win32 Error Codes
	Homepage: https://github.com/rapid7/windows_error
	Path: /home/tmoose/.rvm/gems/ruby-3.0.2/gems/windows_error-0.1.5
	Reverse Dependencies: 
		metasploit-framework (6.2.36) depends on windows_error (>= 0)
		ruby_smb (3.2.2) depends on windows_error (>= 0.1.4)

```

I am running some troubleshooting, but because our princess lives in another castle/gem, it is not super quick...

**Comment by adfoster-r7** [comment]
Looks like this needs a gem bump to get windows_error updated :+1;

**Comment by bwatters-r7** [comment]
Yup; I can access the NTStatus entries, but nothing on the HResult....
Definitely on the 0.1.5
```
msf6 auxiliary(admin/dcerpc/icpr_cert) > irb
[*] Starting IRB shell...
[*] You are in auxiliary/admin/dcerpc/icpr_cert

>> ::WindowsError::NTStatus
=> WindowsError::NTStatus
>> ::Windows::HResult
(irb):2:in `<main>': uninitialized constant Windows (NameError)
Did you mean?  WindowsError                                              
        from /home/tmoose/rapid7/metasploit-framework/lib/rex/ui/text/irb_shell.rb:53:in `block in run'
        from /home/tmoose/rapid7/metasploit-framework/lib/rex/ui/text/irb_shell.rb:52:in `catch'
        from /home/tmoose/rapid7/metasploit-framework/lib/rex/ui/text/irb_shell.rb:52:in `run'
        from /home/tmoose/rapid7/metasploit-framework/lib/msf/ui/console/command_dispatcher/developer.rb:129:in `block in cmd_irb'
        from /home/tmoose/rapid7/metasploit-framework/lib/rex/ui/text/shell/history_manager.rb:49:in `with_context'
        from /home/tmoose/rapid7/metasploit-framework/lib/msf/ui/console/command_dispatcher/developer.rb:125:in `cmd_irb'
        from /home/tmoose/rapid7/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:581:in `run_command'
        from /home/tmoose/rapid7/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:530:in `block in run_single'
        from /home/tmoose/rapid7/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:524:in `each'
        from /home/tmoose/rapid7/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:524:in `run_single'
        from /home/tmoose/rapid7/metasploit-framework/lib/rex/ui/text/shell.rb:162:in `run'
        from /home/tmoose/rapid7/metasploit-framework/lib/metasploit/framework/command/console.rb:48:in `start'
        from /home/tmoose/rapid7/metasploit-framework/lib/metasploit/framework/command/base.rb:82:in `start'
	from ./msfconsole:23:in `<main>'
>> exit
msf6 auxiliary(admin/dcerpc/icpr_cert) > exit
[ruby-3.0.2](land-17499) tmoose@ubuntu:~/rapid7/metasploit-framework$ bundle info windows_error
Your /home/tmoose/.bundle/config config includes `BUNDLE_LOCAL__REX-POWERSHELL`, which contains the dash character (`-`).
This is deprecated, because configuration through `ENV` should be possible, but `ENV` keys cannot include dashes.
Please edit /home/tmoose/.bundle/config and replace any dashes in configuration keys with a triple underscore (`___`).
  * windows_error (0.1.5) 
        Summary: Provides a way to look up Windows NTSTATUS and Win32 Error Codes
        Homepage: https://github.com/rapid7/windows_error
        Path: /home/tmoose/.rvm/gems/ruby-3.0.2/gems/windows_error-0.1.5
        Reverse Dependencies: 
                metasploit-framework (6.2.36) depends on windows_error (>= 0)
                ruby_smb (3.2.2) depends on windows_error (>= 0.1.4)
[ruby-3.0.2](land-17499) tmoose@ubuntu:~/rapid7/metasploit-framework$ 

```

**Comment by zeroSteiner** [comment]
I think I forgot that you have to explicitly require the correct file. I just force-pushed up a change that adds `require 'windows_error/h_result` to the top of the file. That should do the trick for you.

**Comment by bwatters-r7** [comment]
Huh; I'm comfortable saying it works, though I did not get the expected error message, but my setup is not correct for DNS, so I can believe this error....
```
msf6 auxiliary(admin/dcerpc/icpr_cert) > run
[*] Running module against 10.5.134.159

[*] 10.5.134.159:445 - Connecting to ICertPassage (ICPR) Remote Protocol
[*] 10.5.134.159:445 - Binding to \cert...
[+] 10.5.134.159:445 - Bound to \cert
[*] 10.5.134.159:445 - Requesting a certificate for user Administrator - digest algorithm: SHA256 - template: Machine
[-] 10.5.134.159:445 - There was an error while requesting the certificate.
[-] 10.5.134.159:445 - Denied by Policy Module
[-] 10.5.134.159:445 - Error details:
[-] 10.5.134.159:445 -   Source:  (0x0009) FACILITY_SECURITY: The source of the error code is the Security API layer.
[-] 10.5.134.159:445 -   HRESULT: (0x8009480f) CERTSRV_E_SUBJECT_DNS_REQUIRED: The Domain Name System (DNS) name is unavailable and cannot be added to the Subject Alternate name.
[*] Auxiliary module execution completed

```

**Comment by bwatters-r7** [comment]
# Release Notes
This PR pulls in the changes implemented in https://github.com/rapid7/windows_error/pull/5 and adds this new extended error information to the ICPR library code which is used by the icpr_cert module and the new certifried exploit. When the certificate fails to issue for some reason, it looks up the error result as an [HRESULT](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-erref/0642cb2f-2075-4469-918c-4441e69c548a) and prints the details for the user to assist in troubleshooting.

**Pull request merged**